modules/common/src/promises.ts

```diff
@@ -16,3 +16,20 @@ function reflect<T>(prom: Promise<T>) {
     .then((value: T) => ({ value, status: "fulfilled" }))
     .catch((reason: T) => ({ reason, status: "rejected" }));
 }
+
+export async function tryUntil<T>({ func, onFail, shouldContinue }: {
```

Comment:
Still thinking of a better name here... maybe `runWithRetries`, `tryCallback`, `retryFailable`, ...